File: Graph/4.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.parent[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.parent[rootX] = rootY
            else:
                self.parent[rootY] = rootX
                self.rank[rootX] += 1

def kruskal(n, edges, forced_edges=[]):
    uf = UnionFind(n)
    mst_weight = 0
    mst_edges = 0
    selected_edges = []  # لیست یال‌های انتخاب‌شده

    # Add forced edges first
    for u, v, w in forced_edges:
        if uf.find(u) != uf.find(v):
            uf.union(u, v)
            mst_weight += w
            mst_edges += 1
            selected_edges.append((u, v, w))
        else:
            return float('inf'), []
            

    # Sort edges by weight

    # Add remaining edges
    for u, v, w in edges:
        if uf.find(u) != uf.find(v):
            uf.union(u, v)
            mst_weight += w
            mst_edges += 1
            selected_edges.append((u, v, w))
            if mst_edges == n - 1:
                break

    # Check if we formed a valid MST
    if mst_edges == n - 1:
        return mst_weight, selected_edges  # بازگرداندن وزن و یال‌های انتخابی
    else:
        return float('inf'), []  # در صورت عدم امکان تشکیل MST

# Input parsing
n, m = map(int, input().split())
edges = []
for _ in range(m):
    u, v, w = map(int, input().split())
    edges.append((u - 1, v - 1, w))

# edges.sort(key=lambda x: x[2])

s_edge = sorted(edges, key=lambda x: x[2])

q = int(input())
queries = []
for _ in range(q):
    data = list(map(int, input().split()))
    k = data[0]
    queries.append([edges[i - 1] for i in data[1:]])

# Compute the MST weight of the original graph
base_mst_weight = kruskal(n, s_edge)

# Process each query
results = []
for query in queries:
    try:
        mst_weight = kruskal(n, s_edge, query)
        results.append("YES" if mst_weight[0] == base_mst_weight[0] else "NO")
    except:
        results.append("NO")

# Output results
print("\n".join(results))

File: Graph/4.cpp
// https://poe.com/s/gHEbsC9nTR5mpiMsnCSq
// تبدیل کن به CPP
// کد پایتون رو چت جی پیتی زدم سایتش پایین اومده کار نمی کنه به دلیل کمبود وقت ایمیل می کنم براتون

#include <iostream>
#include <vector>
#include <algorithm>
#include <tuple>
#include <climits>  // Include this for INT_MAX

class UnionFind {
public:
    UnionFind(int n) {
        parent.resize(n);
        rank.resize(n, 0);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void unionSets(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            if (rank[rootX] > rank[rootY]) {
                parent[rootY] = rootX;
            } else if (rank[rootX] < rank[rootY]) {
                parent[rootX] = rootY;
            } else {
                parent[rootY] = rootX;
                ++rank[rootX];
            }
        }
    }

private:
    std::vector<int> parent;
    std::vector<int> rank;
};

std::pair<int, std::vector<std::tuple<int, int, int>>> kruskal(int n, std::vector<std::tuple<int, int, int>> edges, std::vector<std::tuple<int, int, int>> forcedEdges = {}) {
    UnionFind uf(n);
    int mstWeight = 0;
    int mstEdges = 0;
    std::vector<std::tuple<int, int, int>> selectedEdges;

    // Add forced edges first
    for (const auto& edge : forcedEdges) {
        int u, v, w;
        std::tie(u, v, w) = edge;
        if (uf.find(u) != uf.find(v)) {
            uf.unionSets(u, v);
            mstWeight += w;
            mstEdges++;
            selectedEdges.push_back(edge);
        } else {
            return {INT_MAX, {}};  // Use INT_MAX from climits
        }
    }

    // Sort edges by weight
    std::sort(edges.begin(), edges.end(), [](const auto& a, const auto& b) {
        return std::get<2>(a) < std::get<2>(b);
    });

    // Add remaining edges
    for (const auto& edge : edges) {
        int u, v, w;
        std::tie(u, v, w) = edge;
        if (uf.find(u) != uf.find(v)) {
            uf.unionSets(u, v);
            mstWeight += w;
            mstEdges++;
            selectedEdges.push_back(edge);
            if (mstEdges == n - 1) {
                break;
            }
        }
    }

    if (mstEdges == n - 1) {
        return {mstWeight, selectedEdges};
    } else {
        return {INT_MAX, {}};  // Use INT_MAX from climits
    }
}

int main() {
    int n, m;
    std::cin >> n >> m;
    std::vector<std::tuple<int, int, int>> edges(m);
    for (int i = 0; i < m; ++i) {
        int u, v, w;
        std::cin >> u >> v >> w;
        edges[i] = {u - 1, v - 1, w};  // Convert to 0-based index
    }


    int q;
    std::cin >> q;
    std::vector<std::vector<std::tuple<int, int, int>>> queries(q);
    for (int i = 0; i < q; ++i) {
        int k;
        std::cin >> k;
        queries[i].resize(k);
        for (int j = 0; j < k; ++j) {
            int idx;
            std::cin >> idx;
            queries[i][j] = edges[idx - 1];  // Convert to 0-based index
        }
    }

    // Sort edges by weight
    std::sort(edges.begin(), edges.end(), [](const auto& a, const auto& b) {
        return std::get<2>(a) < std::get<2>(b);
    });
    // Compute the MST weight of the original graph
    auto baseMST = kruskal(n, edges);

    // Process each query
    std::vector<std::string> results;
    for (const auto& query : queries) {
        auto mstWeight = kruskal(n, edges, query);
        results.push_back(mstWeight.first == baseMST.first ? "YES" : "NO");
    }

    // Output results
    for (const auto& result : results) {
        std::cout << result << std::endl;
    }

    return 0;
}